wagtail/admin/views/generic/models.py

```diff
@@ -919,7 +919,9 @@ def get_context_data(self, **kwargs):
         context["submit_button_label"] = self.submit_button_label
         context["can_delete"] = (
             self.permission_policy is None
-            or self.permission_policy.user_has_permission(self.request.user, "delete")
+            or self.permission_policy.user_has_permission_for_instance(
+                self.request.user, "delete", self.object
+            )
```

Comment:
I agree this is technically the correct method to use, although it makes me wonder how we eventually should do permission checks for the view itself (i.e. the one we currently have via `PermissionCheckedMixin`):

https://github.com/wagtail/wagtail/blob/d742b0278b93660f84f760347d6ba00dada909a6/wagtail/admin/views/generic/permissions.py#L21-L42

If we always use `user_has_permission_for_instance` for detail views (e.g. edit/delete), that means we always fetch the object in question, even if it's not necessary for the check (which is the case for most models).

Not saying that the extra query is a big problem, I just mean the object needs to be fetched at some point before the check happens – in some places we do this in `setup()`, which isn't ideal as discussed in #11590.

In some cases (like collections here) we can do the other way around: use `instances_user_has_permission_for` in `get_queryset`, taking the permissions into account. But this isn't always feasible for more complex application-level checks like ones that are done in `PagePermissionTester`.

---

Makes me wonder how the permissions/actions registry will play into this. Anyway, that's for me to think about I guess, although if you have any immediate thoughts that'd be very welcome 🙂